components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java

```diff
@@ -212,31 +212,46 @@ private IWAAuthenticatedUserBean userInformationInListedUserStores(String authen
                                                                        String userStoreDomains)
             throws AuthenticationFailedException {
 
-        String userStoreDomainForUser = null;
         IWAAuthenticatedUserBean authenticatedUserBean = new IWAAuthenticatedUserBean();
-        try {
-            for (String userStoreDomain : userStoreDomains.split(",")) {
-                resolveUserFromUserStore(
-                        authenticatedUserBean, authenticatedUserName, tenantDomain, userStoreDomain.trim());
-                if (authenticatedUserBean.isUserExists()) {
-                    userStoreDomainForUser = userStoreDomain.trim();
-                    break;
+        authenticatedUserBean.setTenantDomain(tenantDomain);
+        authenticatedUserBean.setUser(authenticatedUserName);
+        MultiAttributeLoginService multiAttributeLoginService =
+                IWAServiceDataHolder.getInstance().getMultiAttributeLoginService();
+
+        if (multiAttributeLoginService.isEnabled(tenantDomain)) {
+            ResolvedUserResult resolvedUserResult = multiAttributeLoginService.resolveUser(
+                    authenticatedUserName, tenantDomain);
+            if (resolvedUserResult != null &&
+                    ResolvedUserResult.UserResolvedStatus.SUCCESS.equals(resolvedUserResult.getResolvedStatus()) &&
+                    userStoreDomains.toUpperCase().contains(resolvedUserResult.getUser().getUserStoreDomain())) {
+                authenticatedUserBean.setUserExists(true);
+                authenticatedUserBean.setUserStoreDomain(resolvedUserResult.getUser().getUserStoreDomain());
+                if (StringUtils.isNotBlank(resolvedUserResult.getUser().getPreferredUsername())) {
+                    authenticatedUserBean.setUser(resolvedUserResult.getUser().getPreferredUsername());
                 }
             }
+        } else {
+            boolean isUserExists = false;
+            String userStoreDomainForUser = null;
+            try {
+                for (String userStoreDomain : userStoreDomains.split(",")) {
+                    if (isUserExistsInUserStore(authenticatedUserName, tenantDomain, userStoreDomain.trim())) {
+                        isUserExists = true;
+                        userStoreDomainForUser = userStoreDomain.trim();
+                        break;
+                    }
+                }
+                authenticatedUserBean.setUserExists(isUserExists);
```

Comment:
Can't we relocate these two lines

authenticatedUserBean.setUserExists(isUserExists);
authenticatedUserBean.setUserStoreDomain(userStoreDomainForUser);

into the if part ? i.e

for (String userStoreDomain : userStoreDomains.split(",")) {
                    if (isUserExistsInUserStore(authenticatedUserName, tenantDomain, userStoreDomain.trim())) {
                        isUserExists = true;
                        userStoreDomainForUser = userStoreDomain.trim();

authenticatedUserBean.setUserExists(true);
authenticatedUserBean.setUserStoreDomain(userStoreDomain.trim());
                        break;
                    }
                }

Reply:
updated with [6f93b37](https://github.com/wso2-extensions/identity-local-auth-iwa-kerberos/pull/77/commits/6f93b37a06ebaf7283992851d1692a50f00853bd)